### PR TITLE
[Dynamo] Support torch.Tensor.fn as TorchVariable, not UserDefinedObjectVariable, preventing graph break

### PR DESCRIPTION
**Pull request by min-jean-cho**
As found in #92709, thanks to @ngimel and @jansel, currently `torch.Tensor.fn` points to `UserDefinedObjectVariable` rather than `TorchVariable`. The root cause is due to https://github.com/pytorch/pytorch/pull/92709#pullrequestreview-1273357406. To prevent this, build `TorchVariable`  of `torch.Tensor.fn` pointing to `torch.ops.aten.fn`.

This issue propagates to `torch.Tensor.fn` causing graph break with `nopython=True`.
```python
import torch
import torch._dynamo as dynamo

#op = torch.ops.aten.abs_ # no graph break
op = torch.Tensor.abs_ # graph break 
args = torch.empty(10)

def foo(args):
    return op(args)

opt_foo = dynamo.optimize("inductor", nopython=True)(foo)
y_ = opt_foo(args)

```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by min-jean-cho** [comment]
Hi @jansel, please help have a look if it makes sense to support `torch.Tensor.{fn}` as `torch.ops.aten.{fn}`, thanks !

File: test/inductor/test_torchinductor_opinfo.py [hunk 1]
```diff
@@ -250,7 +250,8 @@ def process(device_type):
     "stft": {f32, f64},
     "tensor_split": {b8, f16, f32, f64, i32, i64},
     "to_sparse": {f32, f64},
-    "uniform": {f16, f32, f64},
+    # AssertionError: Tensor-likes are not close!
+    "uniform": {f16},
```
Comment (jansel) [inline test/inductor/test_torchinductor_opinfo.py:254]: Interesting that these started working.  I'd expect randomness to not work with opinfo testing.

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -420,6 +420,20 @@ def index_source(key):
                 source=self.source,
                 guards=make_guards(GuardBuilder.FUNCTION_MATCH),
             )
+        elif (
+            not is_allowed(value)
```
Comment (jansel) [inline torch/_dynamo/variables/builder.py:424]: Can we just make `is_allowed()` return True for these?

Should be able to modify:
https://github.com/pytorch/pytorch/blob/74592a43d0d33a6c809fdcfc20249e1c93e7216e/torch/_dynamo/allowed_functions.py#L129

with something like:
```
for name in dir(torch.Tensor):
   method = getattr(torch.Tensor, name)
   if isinstance(method, types.MethodDescriptorType):
       torch_object_ids[id(method)] = f"torch.Tensor.{name}"
```

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/builder.py (hunk 2 above)
Comment (min-jean-cho) [reply torch/_dynamo/variables/builder.py:424]: Hmm, `torch.Tensor.{name}` are method descriptors not module or method, hence doesn't have `__module__` attribute. This issue will propagate to when generating fx graph and finding the module of the method: 

https://github.com/pytorch/pytorch/blob/e7ace1ff930db30deb907c753dc9e41b26f2586f/torch/fx/node.py#L40

File: test/inductor/test_torchinductor_opinfo.py (hunk 1 above)
Comment (min-jean-cho) [reply test/inductor/test_torchinductor_opinfo.py:254]: I'm not sure why eager and inductor randomness match -- not sure if the default generator is global or configured separately and the backend is setting the seed (by the default generator) . 

Here's a simple reproducer:
```
import torch
import torch._dynamo as dynamo

op = torch.Tensor.bernoulli_
args = torch.empty(10000)

def foo(args):
    #torch.manual_seed(42)
    return op(args)

opt_foo = dynamo.optimize("inductor")(foo)

y = foo(args)
y_ = opt_foo(args)
print(y)
print(y_)
print(torch.allclose(y, y_))
```

File: torch/_dynamo/variables/builder.py (hunk 2 above)
Comment (jansel) [reply torch/_dynamo/variables/builder.py:424]: Could we change FX to handle that?  I suspect it just involves adding a case to the code printer.

Comment (min-jean-cho) [reply torch/_dynamo/variables/builder.py]: Thanks @jansel, your suspect was right. Please help have another look.

File: torch/fx/node.py [hunk 3]
```diff
@@ -70,6 +70,9 @@ def _get_qualified_name(func: Callable[..., Any]) -> str:
     # things like getattr just appear in builtins
     if getattr(builtins, func.__name__, None) is func:
         return func.__name__
+    # torch.Tensor.{fn}
+    if isinstance(func, types.MethodDescriptorType) and func.__name__ in dir(torch.Tensor):
```
Comment (jansel) [inline torch/fx/node.py:74]:
```suggestion
    if isinstance(func, types.MethodDescriptorType) and func is getattr(torch.Tensor, func.__name__, None):
```

File: torch/fx/graph.py [hunk 4]
```diff
@@ -514,15 +515,17 @@ def emit_node(node : Node):
...
                     assert isinstance(node.args, tuple)
                     body.append(f'{repr(node)}{maybe_type_annotation} = '
                                 f'{magic_methods[node.target.__name__].format(*(repr(a) for a in node.args))}')
                     return
 
                 # pretty print inplace operators; required for jit.script to work properly
                 # not currently supported in normal FX graphs, but generated by torchdynamo
-                if node.target.__module__ == '_operator' and node.target.__name__ in inplace_methods:
+                if not isinstance(node.target, types.MethodDescriptorType) and \
+                   node.target.__module__ == '_operator' and node.target.__name__ in inplace_methods:
```
Comment (jansel) [inline torch/fx/graph.py:528]:
```suggestion
                if getattr(node.target, "__module__", "") == '_operator' and node.target.__name__ in inplace_methods:
```
Same change above.

**Review by jansel — CHANGES_REQUESTED** [review]

**Review by jansel — APPROVED** [review]

**Comment by min-jean-cho** [comment]
Thanks @jansel, for the help in root cause and review.

File: test/test_torch.py [hunk 5]
```diff
@@ -1525,6 +1525,10 @@ def test_nondeterministic_alert_put(self, device):
                 lambda: op_call(a, indices, values, accumulate=False),
                 'put_')
 
+    # warn_only=False correctly raises RuntimeError: put_ does not have a deterministic implementation
+    # warn_only=True logs warning from the FallbackKernel: torch.ops.aten.put_.default, instead of as UserWarning:
+    # [W Context.cpp:%(lineno)] Warning: put_ does not have a deterministic implementation
+    @skipIfTorchInductor("warning is logged from the FallbackKernel: torch.ops.aten.put_.default when warn_only=True")
```
Comment (min-jean-cho) [inline test/test_torch.py:1531]: Hi @jansel, skipping the following test. Warning when `warn_only=True` is correctly raised form the fallback kernel when graph break is eliminated.

**Comment by PaliC** [comment]
Edit: for Proof of Concept on https://github.com/pytorch/pytorch/pull/94324

## Merge failed
**Reason**: 1 jobs have failed, first few of them are: [trunk / win-vs2019-cuda11.6-py3 / test (functorch, 1, 1, windows.g5.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/4109701114/jobs/7093933409)

<details><summary>Details for Dev Infra team</summary>
<div> Rule `superuser`</div>
<div> Raised by <a href="https://github.com/pytorch/pytorch/actions/runs/4109700512">workflow job</a></div> 
</details>

File: test/inductor/test_torchinductor_opinfo.py (hunk 1 above)
Comment (jansel) [reply test/inductor/test_torchinductor_opinfo.py:254]: Is there a graph break here?  Try changing `dynamo.optimize("inductor", nopython=True)`.

Comment (min-jean-cho) [reply test/inductor/test_torchinductor_opinfo.py:254]: No, there's no graph break with this PR. I could reproduce the randomness match with `nopython=True` as well. I've generally noticed they match for CPU fp32, fp64.

**Pull request closed without merging**